Source/WebCore/page/Navigation.cpp

```diff
@@ -104,18 +105,63 @@ static std::optional<size_t> getEntryIndexOfHistoryItem(const Vector<Ref<Navigat
 }
 
 // https://html.spec.whatwg.org/multipage/nav-history-apis.html#initialize-the-navigation-api-entries-for-a-new-document
-void Navigation::initializeEntries(Ref<HistoryItem>&& currentItem, Vector<Ref<HistoryItem>>& items)
+void Navigation::initializeForNewWindow(std::optional<NavigationNavigationType> navigationType, LocalDOMWindow* previousWindow)
 {
     ASSERT(m_entries.isEmpty());
     ASSERT(!m_currentEntryIndex);
 
     if (hasEntriesAndEventsDisabled())
         return;
 
+    RefPtr page = frame()->page();
+    if (!page)
+        return;
+
+    RefPtr currentItem = frame()->history().currentItem();
+    if (!currentItem)
+        return;
+    if (previousWindow) {
+        Ref previousNavigation = previousWindow->protectedNavigation();
+        if (previousNavigation->m_entries.size() && navigationType && navigationType == NavigationNavigationType::Reload && frame()->document()->protectedSecurityOrigin()->isSameOriginDomain(previousWindow->document()->protectedSecurityOrigin())) {
+            for (auto& entry : previousNavigation->m_entries)
+                m_entries.append(NavigationHistoryEntry::create(protectedScriptExecutionContext().get(), entry.get()));
+            m_currentEntryIndex = previousNavigation->m_currentEntryIndex;
+            updateForActivation(currentItem.get(), navigationType);
+            return;
+        }
+    }
+
+    Vector<Ref<HistoryItem>> items;
+
+    // https://html.spec.whatwg.org/multipage/browsing-the-web.html#getting-session-history-entries-for-the-navigation-api
+    items.append(*currentItem);
+
+    auto rawEntries = page->checkedBackForward()->allItems();
+    auto startingIndex = rawEntries.find(*currentItem);
+    if (startingIndex != notFound) {
+        Ref startingOrigin = SecurityOrigin::create(rawEntries[startingIndex]->url());
+
+        for (int64_t i = startingIndex - 1; i >= 0; i--) {
+            Ref item = rawEntries[i];
+
+            if (!SecurityOrigin::create(item->url())->isSameOriginAs(startingOrigin))
+                break;
+            items.insert(0, WTFMove(item));
```

Comment:
Repeated prepending to a Vector is O(n^2) that can easily be changed to O(n).

Reply:
Good point, should be fixed.